Core/Src/main.c

```diff
@@ -0,0 +1,956 @@
+    uint32_t au32_ticks = (HAL_RCC_GetHCLKFreq() / 1000000);
+    au32_microseconds *= au32_ticks;
+    while ((DWT->CYCCNT - au32_initial_ticks) < au32_microseconds-au32_ticks);
+}
+/* USER CODE END PM */
+
+/* Private variables ---------------------------------------------------------*/
+TIM_HandleTypeDef htim1;
+TIM_HandleTypeDef htim2;
+TIM_HandleTypeDef htim3;
+TIM_HandleTypeDef htim4;
+TIM_HandleTypeDef htim6;
+TIM_HandleTypeDef htim8;
+TIM_HandleTypeDef htim15;
+TIM_HandleTypeDef htim16;
+DMA_HandleTypeDef hdma_tim15_ch1;
+
+/* USER CODE BEGIN PV */
+uint16_t value = TROTTLE_MIN;// throttle value to be sent
+uint16_t ctr = 0;            // arming sequence counter
+uint8_t trottle_down = 0;    // dunnot change trottle up.
+uint8_t armed = 0;           // arming sequence ended
+
+uint32_t motor[DSHOT_FRAME_SIZE];// duty cycles array
+volatile uint16_t DT = 0;
+volatile int8_t motor_new = 0;
+extern USBD_HandleTypeDef hUsbDeviceFS;
+struct Data {
+    int8_t A, B, C, D;
+};
+struct DataA {
+    int8_t addr;
+    int32_t A, B, C;
+};
+/* USER CODE END PV */
+
+/* Private function prototypes -----------------------------------------------*/
+void SystemClock_Config(void);
+static void MX_GPIO_Init(void);
+static void MX_DMA_Init(void);
+static void MX_TIM1_Init(void);
+static void MX_TIM2_Init(void);
+static void MX_TIM16_Init(void);
+static void MX_TIM3_Init(void);
+static void MX_TIM4_Init(void);
+static void MX_TIM8_Init(void);
+static void MX_TIM6_Init(void);
+static void MX_TIM15_Init(void);
+/* USER CODE BEGIN PFP */
+int32_t max(int32_t a, int32_t b);
+int32_t min(int32_t a, int32_t b);
+int32_t clamp(int32_t max, int32_t min, int32_t val);
+void dshot600(uint32_t *motor, uint16_t value);
+
+uint16_t prepareDshotPacket(const uint16_t value);
+
+struct DataA d1 = {2};
+
+int last_error[3], error[3], integral[3], derevative[3];
+int Kp=4000, Ki=100, Kd=0;//TODO: tune coefficients
+
+volatile int8_t motorSpeeds[3];
+
+int32_t dc[3];
+int16_t enc_val_prev[3];
+//float rps_s[3];
+int32_t control[3];
+void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef *htim) {
+    if(htim == &htim6) {
+        int16_t counts_now[3] = {TIM3->CNT, TIM1->CNT, TIM2->CNT};
+        for(char i = 0; i < 3; i++) {
+            dc[i] = counts_now[i] - enc_val_prev[i];
+            enc_val_prev[i] = counts_now[i];
+            int t = ((int)motorSpeeds[i]);
+            error[i] = t - dc[i];
+            integral[i] = clamp(65535/Ki, -65535/Ki, integral[i] + error[i]);
+            derevative[i] = error[i] - last_error[i];
+            last_error[i] = error[i];
+            control[i] = ((Kp * error[i]) + (Ki * integral[i]) + (Kd * derevative[i]));
+
+            control[i] = clamp(65535, -65535, control[i]);
+
+        }
+        d1.A = dc[0];
+        d1.B = (int)motorSpeeds[0];
+        d1.C = error[0];
+        USBD_CUSTOM_HID_SendReport(&hUsbDeviceFS, (uint8_t *) &d1, sizeof(struct DataA));
+
+        //TODO: Not sure about motor order can be changed

```

Comment:
Motor order probably does not need to be changed in firmware. It's possible to connect motor wires differently or change motor order in software.

Reply:
It can be changed because of wires length